sdk/src/main/java/com/tendarts/sdk/client/TDKeysHandler.java

```diff
@@ -0,0 +1,217 @@
+        String data = null;
+
+        try {
+            JSONObject keyJson = new JSONObject();
+            keyJson.put("label", key);
+
+            JSONObject paramsJson = new JSONObject();
+            paramsJson.put("key", keyJson);
+            paramsJson.put("kind", kind.getKindValue());
+            if (value != null) {
+                paramsJson.put("value", value);
+            }
+            paramsJson.put("device", device);
+
+            data = paramsJson.toString();
+
+        } catch (JSONException e) {
+            if (BuildConfig.DEBUG) {
+                Log.d(LOG_TAG, e.getLocalizedMessage());
+            }
+            e.printStackTrace();
+
+            if (listener != null) {
+                listener.onError(e.getLocalizedMessage());
+            }
+        }
+
+        if (data != null) {
+
+            if (BuildConfig.DEBUG) {
+                Log.d(LOG_TAG, "Associate key-value with device: " + data);
+            }
+
+            Communications.postData(Constants.KEYS_DEVICES, Util.getProvider(), 0, new ICommunicationObserver() {
+                @Override
+                public void onSuccess(int operationId, JSONObject data) {
+
+                    String message = "Successfully sent key-value: " + key + "-" + value;
+
+                    TendartsClient.instance(context).logEvent("KEYS", "keyDevice", message);
+
+                    Log.i(LOG_TAG, message);
+
+                    if (listener != null) {
+                        listener.onSuccess();
+                    }
+
+                }
+
+                @Override
+                public void onFail(int operationId, String reason, Communications.PendingCommunication pending) {
+
+                    Util.checkUnauthorized(reason, context);
+
+                    String message = String.format("Error sending key-value: %s-%s. Reason: %s", key, value, reason);
+
+                    TendartsClient.instance(context).logEvent("KEYS", "keyDevice", message);
+
+                    Log.e(LOG_TAG, message);
+
+                    if (listener != null) {
+                        listener.onError(reason);
+                    }
+
+                }
+            }, data);
+
+        }
+
+    }
+
+    public static void associateKeyValueWithUser(final Context context,
+                                                 final String key, KeyValueKind kind, final String value,
+                                                 final TDKeysHandlerInterface listener) {
+
+        if (key == null) {
+            TendartsClient.instance(context).logEvent("KEYS", "keys send error", "Set key");
+        }
+
+        String user = Configuration.instance(context).getUserCode();
+        if( user == null ) {
+            if (listener != null) {
+                listener.onError("The user should be already registered");
+            }
+            return;
+        }
+
+
+        String data = null;
+
+        try {
+            JSONObject keyJson = new JSONObject();
```

Comment:
Can't this be refactorized ? i see this code repeated again and again

Reply:
It's two times in this class, but with different value, do you refer to that?